components/service/experiments/src/main/java/mozilla/components/service/experiments/Experiments.kt

```diff
@@ -47,34 +48,39 @@ open class ExperimentsInternalAPI internal constructor() {
      * @param applicationContext [Context] to access application features, such
      * as shared preferences.
      */
-    fun initialize(
+    open fun initialize(
         applicationContext: Context,
-        fetchClient: Client = HttpURLConnectionClient()
+        configuration: Configuration = Configuration()
     ) {
         if (isInitialized) {
             logger.error("Experiments library should not be initialized multiple times")
             return
         }
 
+        // Any code below might trigger recording into Glean, so make sure Glean is initialized.
+        if (!getGlean().isInitialized()) {
+            logger.error("Glean library must be initialized first")
+            return
+        }
+
         experimentsResult = ExperimentsSnapshot(listOf(), null)
         experimentsLoaded = false
 
-        if (source == null) {
-            source = KintoExperimentSource(
-                EXPERIMENTS_BASE_URL,
-                EXPERIMENTS_BUCKET_NAME,
-                EXPERIMENTS_COLLECTION_NAME,
-                fetchClient
-            )
-        }
-        if (storage == null) {
-            storage = FlatFileExperimentStorage(
-                File(applicationContext.getDir(EXPERIMENTS_DATA_DIR, Context.MODE_PRIVATE),
-                    EXPERIMENTS_JSON_FILENAME)
-            )
-        }
+        storage = getExperimentsStorage(applicationContext)
 
         isInitialized = true
+
+        // Load cached experiments from storage. After this, experiments status
+        // is available.
+        loadExperiments()
+
+        // We now have the last known experiment state loaded for product code
+        // that needs to check it early in startup.
+        // Next we need to update the experiments list from the server async,
+        // without blocking the app launch and schedule future periodic
+        // updates.
+        updater = getExperimentsUpdater(applicationContext)
+        updater!!.initialize(configuration)
```

Comment:
nit: note that this will throw a null-pointer exception in the very unlikely case that `updater` is null at this point. In other places below we're using `updater?.let {}`.

Reply:
With the larger testing refactoring, `updater` & `storage` can actually just be `lateinit`. Changing this.